lib/saluki-components/src/sources/otlp/traces/normalize.rs

```diff
@@ -0,0 +1,643 @@
+    // 2. Replace non-alphanumeric characters (except '.' and '_') with '_'.
+    // 3. Avoid consecutive underscores.
+    // 4. Avoid underscores after periods.
+    // 5. Avoid trailing underscores.
+    let mut normalized = String::with_capacity(name.len());
+    // Skip non-alphabetic characters at start
+    let mut i = 0;
+    let name_bytes = name.as_bytes();
+    while i < name_bytes.len() && !IS_ALPHA_LOOKUP[name_bytes[i] as usize] {
+        i += 1;
+    }
+
+    if i >= name_bytes.len() {
+        return DEFAULT_SPAN_NAME.clone();
+    }
+    if is_valid_metric_name(&name.as_ref()[i..]) {
+        if name.ends_with('_') {
+            return MetaString::from(&name.as_ref()[i..(name.len() - 1)]);
+        }
+        return MetaString::from(&name.as_ref()[i..]);
+    }
+    let mut prev_char = '\0';
+
+    while i < name_bytes.len() {
+        let c = name_bytes[i] as char;
+        i += 1;
+        if (c as u32) < 256 && IS_ALPHA_NUM_LOOKUP[c as usize] {
+            normalized.push(c);
+            prev_char = c;
+        } else if c == '.' {
+            if prev_char == '_' {
+                // Overwrite underscore with period
+                normalized.pop();
+                normalized.push('.');
+                prev_char = '.';
+            } else if prev_char != '.' {
+                normalized.push('.');
+                prev_char = '.';
+            }
+        } else {
+            // Treat as underscore
+            if prev_char != '.' && prev_char != '_' {
+                normalized.push('_');
+                prev_char = '_';
+            }
+        }
+    }
+
+    if normalized.ends_with('_') {
+        normalized.pop();
+    }
+
+    if normalized.is_empty() {
+        return DEFAULT_SPAN_NAME.clone();
+    }
+
+    MetaString::from(normalized)
+}
+
+/// Normalizes a service name.
+///
+/// Truncates to `MAX_SERVICE_LEN` and ensures characters are valid for tags.
+pub fn normalize_service(service: &MetaString) -> MetaString {
+    // TODO: add fall back service for languages
+    // e.g. https://github.com/DataDog/datadog-agent/blob/instrument-otlp-traffic/pkg/trace/traceutil/normalize/normalize.go#L124
+    if service.is_empty() {
+        return DEFAULT_SERVICE_NAME.clone();
+    }
+
+    let truncated = truncate_utf8(service, MAX_SERVICE_LEN);
+    let normalized = normalize_tag_value(truncated);
+
+    if normalized.is_empty() {
+        return DEFAULT_SERVICE_NAME.clone();
+    }
+
+    normalized
+}
+
+/// Normalizes a peer service name.
+///
+/// Returns an empty string if the input is empty or normalizes down to nothing.
+#[allow(dead_code)]
+pub fn normalize_peer_service(service: &MetaString) -> MetaString {
+    if service.is_empty() {
+        return MetaString::from_static("");
+    }
+
+    let truncated = truncate_utf8(service, MAX_SERVICE_LEN);
+    let normalized = normalize_tag_value(truncated);
+
+    if normalized.is_empty() {
+        return MetaString::from_static("");
+    }
+
+    normalized
```

Comment:
You can replace these lines with `normalize_tag_value(truncated)` by itself since it already returns `MetaString`.